storage/isolation_manager.rs

```diff
@@ -241,297 +358,317 @@ struct Timeline {
+        // //             --> However this would definitely require a lock over both
+        // //             --> It should be good enough to order it such that commit/close does: 1) update timeline, then 2) update watermark if possible 3) return control. This will guarantee new snapshots will see this commit in the watermark if it moves up.
+        // //             --> There is a complexity in updating the watermark - we need to scan commit statuses from the current watermark upwards, then update it. What if two commits do this concurrently and race? Is it possible to not make progress?
+        // //
+        // //     We should just re-think what the watermark guarantees actually are...?
+        // //     1) Because we commit optimistically/at the end of a snapshot only, commit slots are only filled once a transaction goes to commit. This means we can at worst be as behind as the current set of 'processing' commits, which is not many.
+        // //     2) We do allow pending commits to proceed out of order sometimes - if they have non-intersecting commit records. This means we can't guarantee the watermark is raised after a commit.
+        // //        --> first hint: maybe we can just bump the watermark when a commit happens that did not skip any commit records/intersected with all concurrent commits? Seems like most cases will not suit this...
+        // //     3) Side note: If we want to preserve causality between transactions by the same user, a new transaction by the same user must wait for the watermark to rise past the recently committed snapshot.
+        // //         --> simplest way to do this without retaining causality information on the client is: when a client asks for a snapshot, immediately get _latest_ (possibly pending) commit sequence number. Then wait until watermark rises to meet that and use it.
+        // //
+        self.watermark.load(Ordering::SeqCst)
     }
 
-    fn record_committed(&self, sequence_number: SequenceNumber) {
-        let window = self.get_or_create_window(sequence_number);
-        window.set_committed(sequence_number);
-        self.remove_window_reader(sequence_number, &window);
+    fn record_reader(&self, sequence_number: SequenceNumber) {
+        if let Some(window) = self.find_window_for_reader(sequence_number) {
+            window.increment_readers();
+        };
     }
 
     fn remove_reader(&self, reader_sequence_number: SequenceNumber) {
-        let read_window = self.get_window(reader_sequence_number);
-        self.remove_window_reader(reader_sequence_number, &read_window);
+        if let Some(window) = self.find_window_for_reader(reader_sequence_number) {
+            debug_assert!(window.get_readers() >= 0);
+            let _readers_remaining = window.decrement_readers();
+            if _readers_remaining == 0 {
+                self.may_free_windows();
+            }
+        };
     }
 
-    fn remove_window_reader(
+    fn find_window_for_reader(
```

Comment:
I've just realised this is actually non-specialised to `reader` right? This is actually just a `try_get_window` by sequence number?
Note that in another place you're doing this operation is in `collect_concurrent_windows` right?